File: reimplementation/adl2/crates/adl-viz/src/label.rs
//! Human-readable labels for HIR nodes, built only from `adl-sema`'s
//! public API. The dump module's renderer in `adl-sema` is crate-private,
//! so the visualizer carries its own; the two share the same source of
//! truth (the typed quantity model) and agree by construction.

use adl_sema::{ArithOp, Fragment, HKind, HNode};
use adl_sema::{Collection, CollectionId, Hir, ParticleRef, Quantity, QuantityArg, ScalarSource};
use adl_syntax::ast::BandKind;

/// Borrows the parts of an `Hir` needed to render any node to text.
pub struct Labeler<'h> {
    hir: &'h Hir,
}

impl<'h> Labeler<'h> {
    #[must_use]
    pub fn new(hir: &'h Hir) -> Self {
        Self { hir }
    }

    /// A collection's first-bound name, or its base name, falling back to
    /// the structural id. Never panics on out-of-range ids.
    #[must_use]
    pub fn collection(&self, id: CollectionId) -> String {
        if let Some(name) = self
            .hir
            .coll_names
            .get(id.0 as usize)
            .and_then(|names| names.first())
        {
            return self.hir.symbols.display(*name).to_owned();
        }
        match self.hir.table.collection(id) {
            Collection::Base(sym) => self.hir.symbols.display(*sym).to_owned(),
            Collection::Filtered { parent, .. } => format!("filter({})", self.collection(*parent)),
            Collection::Union(parts) => {
                let parts: Vec<String> = parts.iter().map(|&p| self.collection(p)).collect();
                format!("union({})", parts.join(", "))
            }
            Collection::Combination { parts, .. } => {
                let parts: Vec<String> = parts.iter().map(|&p| self.collection(p)).collect();
                format!("comb({})", parts.join(", "))
            }
            Collection::Sorted { source, .. } => format!("sort({})", self.collection(*source)),
            Collection::Slice { source, start, end } => match end {
                Some(e) => format!("{}[{start}:{e}]", self.collection(*source)),
                None => format!("{}[{start}:]", self.collection(*source)),
            },
            Collection::CombProject { comb, axis } => {
                let field = match axis {
                    adl_sema::CombAxis::Member(s) | adl_sema::CombAxis::Candidate(s) => {
                        self.hir.symbols.display(*s).to_owned()
                    }
                };
                format!("{}->{field}", self.collection(*comb))
            }
        }
    }

    fn particle(&self, p: &ParticleRef) -> String {
        match p {
            ParticleRef::Elem { coll, index } => format!("{}[{index}]", self.collection(*coll)),
            ParticleRef::Whole(coll) => format!("{}[*]", self.collection(*coll)),
            ParticleRef::Met => "MET".to_owned(),
            ParticleRef::Binder { coll, name } => {
                format!(
                    "{}@{}",
                    self.collection(*coll),
                    self.hir.symbols.display(*name)
                )
            }
            ParticleRef::ThisElem => "this".to_owned(),
            ParticleRef::ReduceElem => "elem".to_owned(),
            ParticleRef::Sum(parts) => {
                let parts: Vec<String> = parts.iter().map(|p| self.particle(p)).collect();
                format!("({})", parts.join(" + "))
            }
        }
    }

    fn quantity(&self, q: &Quantity) -> String {
        let t = &self.hir.table;
        match q {
            Quantity::EventScalar(src) => match src {
                ScalarSource::MetProp(p) => format!("MET.{}", t.prop_display(*p)),
                ScalarSource::EventVar(s) => format!("evt.{}", self.hir.symbols.display(*s)),
                ScalarSource::Trigger(s) => format!("trig({})", self.hir.symbols.display(*s)),
            },
            Quantity::Size(c) => format!("size({})", self.collection(*c)),
            Quantity::ElemProp { coll, index, prop } => {
                format!(
                    "{}[{index}].{}",
                    self.collection(*coll),
                    t.prop_display(*prop)
                )
            }
            Quantity::AngularSep { kind, a, b, .. } => {
                format!(
                    "{}({}, {})",
                    kind.as_str(),
                    self.particle(a),
                    self.particle(b)
                )
            }
            Quantity::ExternalFn { name, args } => {
                let args: Vec<String> = args.iter().map(|a| self.arg(a)).collect();
                format!("{}({})", self.hir.symbols.display(*name), args.join(", "))
            }
        }
    }

    fn arg(&self, a: &QuantityArg) -> String {
        let t = &self.hir.table;
        match a {
            QuantityArg::Num(n) => n.clone(),
            // Opaque args are sema-interned text keyed on `C<id>#name`
            // collection spellings; strip the id prefix for display (the
            // identity it encodes is unchanged — this is label-only).
            QuantityArg::Opaque(n) => strip_coll_ids(n),
            QuantityArg::Quantity(q) => self.quantity(t.quantity(*q)),
            QuantityArg::Particle(p) => self.particle(p),
            QuantityArg::Collection(c) => self.collection(*c),
            QuantityArg::CollProp { coll, prop } => {
                format!("{}[*].{}", self.collection(*coll), t.prop_display(*prop))
            }
        }
    }

    /// Render an HIR expression node to a single-line label.
    #[must_use]
    pub fn node(&self, n: &HNode) -> String {
        let t = &self.hir.table;
        match &n.kind {
            HKind::Num(s) => s.clone(),
            HKind::Bool(b) => b.to_string(),
            HKind::Quantity(q) => self.quantity(t.quantity(*q)),
            HKind::ElemSelfProp(p) => format!("this.{}", t.prop_display(*p)),
            HKind::ReduceProp(p) => format!("elem.{}", t.prop_display(*p)),
            HKind::Reduce { kind, coll, body, .. } => {
                format!("{}({}: {})", kind.as_str(), self.collection(*coll), self.node(body))
            }
            HKind::ScalarMinMax { kind, args } => {
                let inner: Vec<String> = args.iter().map(|a| self.node(a)).collect();
                format!("{}({})", kind.as_str(), inner.join(", "))
            }
            HKind::CollProp { coll, prop } => {
                format!("{}[*].{}", self.collection(*coll), t.prop_display(*prop))
            }
            HKind::Particle(p) => self.particle(p),
            HKind::CollValue(c) => self.collection(*c),
            HKind::Neg(e) => format!("(- {})", self.node(e)),
            HKind::Not(e) => format!("(not {})", self.node(e)),
            HKind::Binary { op, lhs, rhs } => {
                format!("({} {} {})", self.node(lhs), op_str(*op), self.node(rhs))
            }
            HKind::And(v) => self.joined(v, " and "),
            HKind::Or(v) => self.joined(v, " or "),
            HKind::Cmp { op, lhs, rhs } => {
                format!("({} {} {})", self.node(lhs), op.as_str(), self.node(rhs))
            }
            HKind::Band { kind, expr, lo, hi } => {
                let op = match kind {
                    BandKind::In => "[]",
                    BandKind::Out => "][",
                };
                format!("({} {op} {lo} {hi})", self.node(expr))
            }
            HKind::Ternary { guard, then, els } => match els {
                Some(e) => format!(
                    "({} ? {} : {})",
                    self.node(guard),
                    self.node(then),
                    self.node(e)
                ),
                None => format!("({} ? {} : true)", self.node(guard), self.node(then)),
            },
            HKind::Abs(e) => format!("abs({})", self.node(e)),
            HKind::RegionPred(i) => {
                let name = self
                    .hir
                    .region_name_order
                    .get(*i)
                    .map_or("?", |&s| self.hir.symbols.display(s));
                format!("region:{name}")
            }
            HKind::Unsupported => match &n.tag {
                Fragment::Unsupported(reason) => format!("<unsupported: {reason}>"),
                Fragment::InFragment => "<unsupported>".to_owned(),
            },
        }
    }

    fn joined(&self, v: &[HNode], sep: &str) -> String {
        let parts: Vec<String> = v.iter().map(|n| self.node(n)).collect();
        format!("({})", parts.join(sep))
    }
}

fn op_str(op: ArithOp) -> &'static str {
    op.as_str()
}

/// Strip `C<digits>#` collection-id prefixes from sema-interned opaque
/// text, leaving the human collection name. `C4#leptons[0]` → `leptons[0]`;
/// non-prefix `#` and identifiers like `MET` are untouched.
fn strip_coll_ids(s: &str) -> String {
    let chars: Vec<char> = s.chars().collect();
    let mut out = String::with_capacity(s.len());
    let mut i = 0;
    while i < chars.len() {
        // A prefix starts at a word boundary with `C`, runs over digits,
        // and must hit `#` to be a collection id.
        let boundary = i == 0 || !chars[i - 1].is_ascii_alphanumeric();
        if boundary && chars[i] == 'C' {
            let mut j = i + 1;
            while j < chars.len() && chars[j].is_ascii_digit() {
                j += 1;
            }
            if j > i + 1 && j < chars.len() && chars[j] == '#' {
                i = j + 1; // skip `C<digits>#`, keep the name that follows
                continue;
            }
        }
        out.push(chars[i]);
        i += 1;
    }
    out
}

#[cfg(test)]
mod tests {
    use super::strip_coll_ids;

    #[test]
    fn strips_collection_ids() {
        assert_eq!(strip_coll_ids("C4#leptons[0].pt"), "leptons[0].pt");
        assert_eq!(
            strip_coll_ids("(C0#jets[0] C0#jets[1])"),
            "(jets[0] jets[1])"
        );
        // No prefix: untouched.
        assert_eq!(strip_coll_ids("MET.phi"), "MET.phi");
        assert_eq!(strip_coll_ids("Cology"), "Cology");
        assert_eq!(strip_coll_ids("foo#bar"), "foo#bar");
    }
}

File: cpp/libs/viz/src/label.cpp
#include "label.hpp"

#include "adl2/sema/ops.hpp"
#include "adl2/sema/quantity.hpp"

#include <string>
#include <vector>

namespace adl2::viz {
namespace {

bool is_ascii_alnum(char c) {
  return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
}

bool is_ascii_digit(char c) { return c >= '0' && c <= '9'; }

std::string join_parts(const std::vector<std::string>& parts, const char* sep) {
  std::string out;
  for (std::size_t i = 0; i < parts.size(); ++i) {
    if (i) out += sep;
    out += parts[i];
  }
  return out;
}

}  // namespace

std::string strip_coll_ids(const std::string& s) {
  std::string out;
  out.reserve(s.size());
  std::size_t i = 0;
  while (i < s.size()) {
    const bool boundary = i == 0 || !is_ascii_alnum(s[i - 1]);
    if (boundary && s[i] == 'C') {
      std::size_t j = i + 1;
      while (j < s.size() && is_ascii_digit(s[j])) ++j;
      if (j > i + 1 && j < s.size() && s[j] == '#') {
        i = j + 1;  // skip `C<digits>#`
        continue;
      }
    }
    out.push_back(s[i]);
    ++i;
  }
  return out;
}

std::string Labeler::collection(adl2::sema::CollectionId id) const {
  const auto& hir = *hir_;
  if (id.id < hir.coll_names.size() && !hir.coll_names[id.id].empty()) {
    return hir.symbols.display(hir.coll_names[id.id].front());
  }
  const auto& c = hir.table.collection(id);
  switch (c.kind) {
    case adl2::sema::CollectionKind::Base:
      return hir.symbols.display(c.base);
    case adl2::sema::CollectionKind::Filtered:
      return "filter(" + collection(c.parent) + ")";
    case adl2::sema::CollectionKind::Union: {
      std::vector<std::string> parts;
      parts.reserve(c.parts.size());
      for (auto p : c.parts) parts.push_back(collection(p));
      return "union(" + join_parts(parts, ", ") + ")";
    }
    case adl2::sema::CollectionKind::Combination: {
      std::vector<std::string> parts;
      parts.reserve(c.parts.size());
      for (auto p : c.parts) parts.push_back(collection(p));
      return "comb(" + join_parts(parts, ", ") + ")";
    }
    case adl2::sema::CollectionKind::Sorted:
      return "sort(" + collection(c.parent) + ")";
    case adl2::sema::CollectionKind::Slice: {
      std::string out = collection(c.parent) + "[" + std::to_string(c.slice_start) + ":";
      if (c.slice_end) out += std::to_string(*c.slice_end);
      out += "]";
      return out;
    }
    case adl2::sema::CollectionKind::CombProject:
      return collection(c.parent) + "->" + hir.symbols.display(c.axis.name);
  }
  return "?";
}

std::string Labeler::particle(const adl2::sema::ParticleRef& p) const {
  using adl2::sema::ParticleKind;
  switch (p.kind) {
    case ParticleKind::Elem:
      return collection(p.coll) + "[" + p.index.to_string() + "]";
    case ParticleKind::Whole:
      return collection(p.coll) + "[*]";
    case ParticleKind::Met:
      return "MET";
    case ParticleKind::Binder:
      return collection(p.coll) + "@" + hir_->symbols.display(p.name);
    case ParticleKind::ThisElem:
      return "this";
    case ParticleKind::ReduceElem:
      return "elem";
    case ParticleKind::Sum: {
      std::vector<std::string> parts;
      parts.reserve(p.parts.size());
      for (const auto& part : p.parts) parts.push_back(particle(part));
      return "(" + join_parts(parts, " + ") + ")";
    }
  }
  return "?";
}

std::string Labeler::quantity(const adl2::sema::Quantity& q) const {
  using adl2::sema::QuantityKind;
  using adl2::sema::ScalarSourceKind;
  const auto& t = hir_->table;
  switch (q.kind) {
    case QuantityKind::EventScalar:
      switch (q.scalar.kind) {
        case ScalarSourceKind::MetProp:
          return "MET." + t.prop_display(q.scalar.prop);
        case ScalarSourceKind::EventVar:
          return "evt." + hir_->symbols.display(q.scalar.name);
        case ScalarSourceKind::Trigger:
          return "trig(" + hir_->symbols.display(q.scalar.name) + ")";
      }
      break;
    case QuantityKind::Size:
      return "size(" + collection(q.coll) + ")";
    case QuantityKind::ElemProp:
      return collection(q.coll) + "[" + q.index.to_string() + "]." + t.prop_display(q.prop);
    case QuantityKind::AngularSep:
      return std::string(adl2::sema::ang_kind_str(q.ang)) + "(" + particle(q.a) + ", " +
             particle(q.b) + ")";
    case QuantityKind::ExternalFn: {
      std::vector<std::string> args;
      args.reserve(q.args.size());
      for (const auto& a : q.args) args.push_back(arg(a));
      return hir_->symbols.display(q.name) + "(" + join_parts(args, ", ") + ")";
    }
    case QuantityKind::Present:
      return "defined(" + quantity(t.quantity(q.inner)) + ")";
  }
  return "?";
}

std::string Labeler::arg(const adl2::sema::QuantityArg& a) const {
  using adl2::sema::QuantityArgKind;
  const auto& t = hir_->table;
  switch (a.kind) {
    case QuantityArgKind::Num:
      return a.text;
    case QuantityArgKind::Opaque:
      return strip_coll_ids(a.text);
    case QuantityArgKind::Quantity:
      return quantity(t.quantity(a.qid));
    case QuantityArgKind::Particle:
      return particle(a.particle);
    case QuantityArgKind::Collection:
      return collection(a.coll);
    case QuantityArgKind::CollProp:
      return collection(a.coll) + "[*]." + t.prop_display(a.prop);
  }
  return "?";
}

std::string Labeler::node(const adl2::sema::HNode& n) const {
  using Kind = adl2::sema::HNode::Kind;
  const auto& t = hir_->table;
  switch (n.kind) {
    case Kind::Num:
      return n.text;
    case Kind::Bool:
      return n.bool_val ? "true" : "false";
    case Kind::Quantity:
      return quantity(t.quantity(n.qid));
    case Kind::ElemSelfProp:
      return "this." + t.prop_display(n.prop);
    case Kind::ReduceProp:
      return "elem." + t.prop_display(n.prop);
    case Kind::Reduce:
      return std::string(adl2::sema::reduce_kind_str(n.reduce)) + "(" + collection(n.coll) +
             ": " + (n.a ? node(*n.a) : "?") + ")";
    case Kind::ScalarMinMax: {
      std::vector<std::string> inner;
      inner.reserve(n.items.size());
      for (const auto& a : n.items) inner.push_back(node(a));
      return std::string(adl2::sema::reduce_kind_str(n.reduce)) + "(" +
             join_parts(inner, ", ") + ")";
    }
    case Kind::CollProp:
      return collection(n.coll) + "[*]." + t.prop_display(n.prop);
    case Kind::Particle:
      return particle(n.particle);
    case Kind::CollValue:
      return collection(n.coll);
    case Kind::Neg:
      return "(- " + (n.a ? node(*n.a) : "?") + ")";
    case Kind::Not:
      return "(not " + (n.a ? node(*n.a) : "?") + ")";
    case Kind::Binary:
      return "(" + (n.a ? node(*n.a) : "?") + " " + adl2::sema::arith_op_str(n.arith) + " " +
             (n.b ? node(*n.b) : "?") + ")";
    case Kind::And:
      return joined(n.items, " and ");
    case Kind::Or:
      return joined(n.items, " or ");
    case Kind::Cmp:
      return "(" + (n.a ? node(*n.a) : "?") + " " + adl2::sema::cmp_op_str(n.cmp) + " " +
             (n.b ? node(*n.b) : "?") + ")";
    case Kind::Band: {
      const char* op = n.band == adl2::sema::BandKind::In ? "[]" : "][";
      return "(" + (n.a ? node(*n.a) : "?") + " " + op + " " + n.lo + " " + n.hi + ")";
    }
    case Kind::Ternary:
      if (n.c) {
        return "(" + (n.a ? node(*n.a) : "?") + " ? " + (n.b ? node(*n.b) : "?") + " : " +
               node(*n.c) + ")";
      }
      return "(" + (n.a ? node(*n.a) : "?") + " ? " + (n.b ? node(*n.b) : "?") + " : true)";
    case Kind::Abs:
      return "abs(" + (n.a ? node(*n.a) : "?") + ")";
    case Kind::RegionPred: {
      std::string name = "?";
      if (n.region_index < hir_->region_name_order.size()) {
        name = hir_->symbols.display(hir_->region_name_order[n.region_index]);
      }
      return "region:" + name;
    }
    case Kind::Unsupported:
      if (!n.tag.is_in_fragment()) {
        return "<unsupported: " + n.tag.reason + ">";
      }
      return "<unsupported>";
  }
  return "?";
}

std::string Labeler::joined(const std::vector<adl2::sema::HNode>& v, const char* sep) const {
  std::vector<std::string> parts;
  parts.reserve(v.size());
  for (const auto& n : v) parts.push_back(node(n));
  return "(" + join_parts(parts, sep) + ")";
}

}  // namespace adl2::viz
